File: src/FinHub/FinHub.Application/Modules/Budgeting/Services/BudgetingService.cs
using FinHub.Application.Common.Interfaces;
using FinHub.Application.Modules.Budgeting.DTOs;
using FinHub.Domain.Entities;

namespace FinHub.Application.Modules.Budgeting.Services;

public sealed class BudgetingService : IBudgetingService
{
    private readonly ICustomerRepository _customerRepository;
    private readonly IBankAccountRepository _bankAccountRepository;
    private readonly IBudgetSnapshotRepository _budgetSnapshotRepository;

    public BudgetingService(
        ICustomerRepository customerRepository,
        IBankAccountRepository bankAccountRepository,
        IBudgetSnapshotRepository budgetSnapshotRepository)
    {
        _customerRepository = customerRepository;
        _bankAccountRepository = bankAccountRepository;
        _budgetSnapshotRepository = budgetSnapshotRepository;
    }

    public async Task<BudgetOverviewDto> GetCustomerBudgetOverviewAsync(Guid customerId, CancellationToken cancellationToken = default)
    {
        var customer = await _customerRepository.GetByIdAsync(customerId, cancellationToken)
            ?? throw new KeyNotFoundException($"Customer with ID '{customerId}' was not found.");

        var accounts = await _bankAccountRepository.GetByCustomerIdAsync(customerId, cancellationToken);
        var totalBalance = accounts.Sum(a => a.Balance.Amount);
        var monthlyBudget = Math.Max(totalBalance, 5000m);
        var monthlySpent = Math.Max(totalBalance * 0.32m, 100m);
        var remainingBudget = monthlyBudget - monthlySpent;
        var savingsGoal = Math.Max(monthlyBudget * 0.15m, 250m);
        var month = DateTimeOffset.UtcNow;

        var snapshot = BudgetSnapshot.Create(
            customer.Id,
            month,
            monthlyBudget,
            monthlySpent,
            remainingBudget,
            savingsGoal,
            "SAR");

        await _budgetSnapshotRepository.AddAsync(snapshot, cancellationToken);

        return new BudgetOverviewDto
        {
            CustomerId = customer.Id,
            Month = month,
            MonthlyBudget = monthlyBudget,
            MonthlySpent = monthlySpent,
            RemainingBudget = remainingBudget,
            SavingsGoal = savingsGoal,
            Currency = "SAR",
            Alerts = new List<BudgetAlertDto>
            {
                new()
                {
                    Type = "SpendingLimit",
                    Title = "Monthly spending alert",
                    Message = "You are tracking month-to-date budget usage.",
                    Threshold = monthlyBudget * 0.8m,
                    IsActive = true
                }
            }
        };
    }
}

File: src/Web/src/app/core/budgeting/budgeting.service.ts
import { Injectable, inject } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';

export interface BudgetAlertDto {
    type: string;
    title: string;
    message: string;
    threshold: number;
    isActive: boolean;
}

export interface BudgetOverviewDto {
    customerId: string;
    month: string;
    monthlyBudget: number;
    monthlySpent: number;
    remainingBudget: number;
    savingsGoal: number;
    currency: string;
    alerts: BudgetAlertDto[];
}

@Injectable({
    providedIn: 'root'
})
export class BudgetingService {
    private http = inject(HttpClient);

    private get API_URL(): string {
        if (typeof window !== 'undefined' && window.location.hostname !== 'localhost' && window.location.hostname !== '127.0.0.1') {
            return 'https://finhub-w9xq.onrender.com/api/v1/budgets';
        }

        return 'http://localhost:5000/api/v1/budgets';
    }

    public getCustomerBudgetOverview(customerId: string): Observable<BudgetOverviewDto> {
        return this.http.get<BudgetOverviewDto>(`${this.API_URL}/customer/${customerId}`);
    }
}
